### ISSUE TITLE
Absolute positioned input device support

### ISSUE THREAD

**Issue by X547** [LekKit/RVVM#58]
Currently RVVM support only relative-positioned PS/2 mouse emulation that require pointer capture. It is annoying and limiting because it do not work properly on touchscreens and tablets with digital pen. TinyEMU and QEmu supports absolute positioned pointer.

It can be implemented with i2c HID or USB HID.

**Comment by LekKit** [comment]
Will be implemented at some point (Already have that in TODO), I also want this functionality. Although there is some work to be done that has higher priorities.
I'm afraid this will not be covered in RVVM 0.5 as this release is already taking too long & would bring more regressions before we can fix existing ones.

For reference, I'll likely implement OpenCores I2C (used in HiFive boards) and later we can connect an i2c touchpad to it.

**Comment by X547** [comment]
There are also touchscreen PS/2 devices exist, example: https://github.com/torvalds/linux/blob/master/drivers/input/mouse/lifebook.c, https://github.com/torvalds/linux/blob/master/drivers/input/mouse/touchkit_ps2.c.

For Haiku I consider this problem a high priority because it is hard to control GUI without absolute pointer positioning or pointer grab support.

**Comment by LekKit** [comment]
> There are also touchscreen PS/2 devices exist, example: https://github.com/torvalds/linux/blob/master/drivers/input/mouse/lifebook.c, https://github.com/torvalds/linux/blob/master/drivers/input/mouse/touchkit_ps2.c.

We considered them at some point, then it turned out that X Server removed support for ps2 touchscreens and the entire idea was scrapped (All focus was on linux guests)

I can implement pointer grabbing as initial effort. I unfortunately have very little progress on new HID devices as of now.

**Comment by X547** [comment]
I may try to implement Opencores i2c host controller and i2c HID in RVVM. It will be also beneficial for improving Haiku laptop touchpad and tablet support. QEmu seems have no i2c HID support.

**Comment by LekKit** [comment]
> I may try to implement Opencores i2c host controller and i2c HID in RVVM. It will be also beneficial for improving Haiku laptop touchpad and tablet support. QEmu seems have no i2c HID support.

Can you leave i2c controller to me? I already have some work on that and knowledge of SiFive doc

If you have API suggestion tho, and want to write actual input devices, then youre good to go. I just have some more ideas concerning that I2C controller...

**Comment by LekKit** [comment]
Built a Linux kernel with OpenCores I2C, I2C HID support for testing. There are also drivers for SiFive GPIO/UART to complement the Unmatched board IO devices (Will be implemented soon).
[linux_6.0.zip](https://github.com/LekKit/RVVM/files/10111457/linux_6.0.zip)
[config_6.0.10.txt](https://github.com/LekKit/RVVM/files/10111478/config_6.0.10.txt)

I already have a working i2c-oc controller prototype
```
┌─ [root@rvvm-arch]  ⚞ UwU ⚟ 
└─╼ ~ $ i2cdetect -l
i2c-0   i2c             i2c-ocores                              I2C adapter
```

Will be back in about an hour or two to discuss I2C API.
Perhaps will be sth like this (Pretty similar to existing PCI bus API but for serial devices), but I'm not sure and open for suggestions
```c
i2c_bus_t* i2c_bus = i2c_oc_init_auto(machine);
i2c_dev_desc_t i2c_desc = {
    .addr = 0x41,           // Slave address on the bus
    .write = i2c_hid_write, // Callback to write from I2C master to slave
};
i2c_dev_t* i2c_dev = i2c_bus_attach(i2c_bus);
ic2_write(i2c_dev, data, size); // Write from I2C slave to master
```

**Comment by X547** [comment]
> Can you leave i2c controller to me? I already have some work on that and knowledge of SiFive doc

How much time it will take to implement for you?

**Comment by LekKit** [comment]
> > Can you leave i2c controller to me? I already have some work on that and knowledge of SiFive doc
> 
> How much time it will take to implement for you?

Probably till the end of this day x)
I don't want to rush with API decisions tho, see prev message (wanna hear your thoughts).

To be fair I'm also thinking about some kind of "common serial device API" or sth, to deal with all uart/i2c/ps2/spi/etc buses where we simply have a bidirectional flow of bytes. Is this a good idea at all?
Currently, ps2 API is a mess (probably not worth touching at this point), and uart is nailed to stdio/stdout. Further i2c/spi devices shouldn't have this kind of rushed approach.

**Comment by X547** [comment]
> To be fair I'm also thinking about some kind of "common serial device API" or sth, to deal with all uart/i2c/ps2/spi/etc buses where we simply have a bidirectional flow of bytes. Is this a good idea at all?

i2c bus have a concept of device addresses and transactions that may consist of multiple read/write operations and different device addresses in single transaction. Device may NACK write operation. With some common serial API it will be needed at least mark end of transaction (i2c stop condition).

Very common i2c bus transaction pattern is a combination of zero/one write and zero/one read (something like `void i2c_op(uint16_t adr, uint8_t *writeData, size_t writeLen, uint8_t *readData, size_t readLen);`).

**Comment by LekKit** [comment]
> > To be fair I'm also thinking about some kind of "common serial device API" or sth, to deal with all uart/i2c/ps2/spi/etc buses where we simply have a bidirectional flow of bytes. Is this a good idea at all?
> 
> i2c bus have a concept of device addresses and transactions that may consist of multiple read/write operations and different device addresses in single transaction. Device may NACK write operation.
> 
> Very common i2c bus transaction pattern is a combination of zero/one write and zero/one read (something like `void i2c_op(uint16_t adr, uint8_t *writeData, size_t writeLen, uint8_t *readData, size_t readLen);`).

Address of the device will map internally in the I2C controller onto different slave device handles (Which are separate serial devices at that point).
NACK may be indicated by return value of the write callback (Similar thing already done in ps2).
Since reads from slaves are always initiated by master we can implement both read/write in device description. I assume the HID devices would be polling-based instead of being IRQ-based then?

**Comment by X547** [comment]
> Since reads from slaves are always initiated by master we can implement both read/write in device description.

How write/read operations can be done in single transaction?

> I assume the HID devices would be polling-based instead of being IRQ-based then?

I2c HID devices define interrupt number in FDT definition that is used to indicate that input packet is ready to read.

https://www.kernel.org/doc/Documentation/devicetree/bindings/input/hid-over-i2c.txt:

> Required properties:
> - compatible: must be "hid-over-i2c"
> - reg: i2c slave address
> - hid-descr-addr: HID descriptor address
> - interrupts: interrupt line

**Comment by X547** [comment]
I first planned API like this:
``` C
struct i2c_device {
    int (*do_io)(void *dev, bool isStart, bool isStop, bool isWrite, uint8_t *data, size_t dataLen); // return count of ACKed bytes
};
```

Per byte:
``` C
struct i2c_device {
    int (*do_io)(void *dev, bool isStart, bool isStop, bool isWrite, uint8_t *data); // return ACK if write
};
```

**Comment by LekKit** [comment]
> How write/read operations can be done in single transaction?

I'll take your approach then, it's simply that a single function with a ton of arguments doesn't look pretty to me. It's possible to use separate start/stop/read/write calls for a device perhaps, but it's really a matter of preference at this point.

> I2c HID devices define interrupt number in FDT definition that is used to indicate that input packet is ready to read.

Oh that's good to know, thanks. Does it use a phandle of the master i2c controller as well, or other methods of describing which bus owns this device?

**Comment by X547** [comment]
> Does it use a phandle of the master i2c controller as well, or other methods of describing which bus owns this device?

I2c host controller is parent FDT node if I am correct.

**Comment by X547** [comment]
> I'll take your approach then, it's simply that a single function with a ton of arguments doesn't look pretty to me.

I do not insist on one function with many flags. It just should be possible to declare i2c transaction bounds. That flags are inspired by Opencores i2c command register flags.

**Comment by LekKit** [comment]
> > I'll take your approach then, it's simply that a single function with a ton of arguments doesn't look pretty to me.
> 
> I do not insist on one function with many flags. It just should be possible to declare i2c transaction bounds. That flags are inspired by Opencores i2c command register flags.

OpenCores I2C also has "repeated start" feature, so that it can change read/write direction or current slave device without stopping the transaction. How should we handle that?

**Comment by X547** [comment]
> How should we handle that?

Just set start flag again without setting stop flag. Set stop flag only once when transaction is finished.

Example (pseudocode): startWrite(); write(1); write(2); write(3); /* no stop here */ startRead(); read(val); read(val); stop();

**Comment by LekKit** [comment]
> > How should we handle that?
> 
> Just set start flag again without setting stop flag. Set stop flag only once when transaction is finished.
> 
> Example (pseudocode): startWrite(); write(1); write(2); write(3); startRead(); read(val); read(val); stop();

Is it a good idea to combine the i2c_op flags into a bitfield (same as command register in OC I2C)?
I.e. `i2c_op(I2C_START | I2C_WR | I2C_STOP, buff, size)` for a write transaction in a single call, etc

**Comment by X547** [comment]
> Is it a good idea to combine the i2c_op flags into a bitfield (same as command register in OC I2C)?

Yes.

**Comment by X547** [comment]
I2c device interface in QEmu: https://github.com/qemu/qemu/blob/d495e432c04a6394126c35cf96517749708b410f/include/hw/i2c/i2c.h#L26.

**Comment by LekKit** [comment]
> I2c device interface in QEmu: https://github.com/qemu/qemu/blob/d495e432c04a6394126c35cf96517749708b410f/include/hw/i2c/i2c.h#L26.

This indeed looks similar to how i initially wanted to implement this, with explicit transfer start/stop and separate read/write calls. Is this fine for you?

**Comment by X547** [comment]
> Is this fine for you?

I don't care as long it is possible to group multiple read/write operations in single transactions and ACK handling. i2c HID often use write/read sequence in single transaction.

It is mostly a matter of taste.

**Comment by LekKit** [comment]
See c9af1c0 for initial OpenCores I2C implementation
I made a simple byte-bang logger device, then used i2c-tools to test if it's actually working - seems fine.

Any suggestions are welcome

**Comment by X547** [comment]
> //void (*start)(); // Not needed since r/w implicitly starts a transaction?

Start means single read/write group in transaction, not only start of transaction, so it is needed. Start is sent when read/write direction changes. Start event includes information about will it be read or write group.

**Comment by LekKit** [comment]
> > //void (*start)(); // Not needed since r/w implicitly starts a transaction?
> 
> Start means single read/write group in transaction, not only start of transaction, so it is needed. Start is sent when read/write direction changes. Start event includes information about will it be read or write group.

Don't the actual r/w calls give enough information tho? I mean, after stop() we surely know any read/write call is a new transaction, and when write happens after read() it's also obvious.
Anyways gonna implement it now, any device that doesn't need it may initialize the callback to NULL. I'm just trying to understand what did I miss.

**Comment by X547** [comment]
Patch:
``` diff
diff --git a/src/devices/i2c-oc.c b/src/devices/i2c-oc.c
index 44e14cb..9cc2d9e 100644
--- a/src/devices/i2c-oc.c
+++ b/src/devices/i2c-oc.c
@@ -141,8 +141,11 @@ static bool i2c_oc_mmio_write(rvvm_mmio_dev_t* dev, void* data, size_t offset, u
             if ((cmd & I2C_OC_CR_WR)) {
                 if (bus->sel_addr == 0xFFFF) {
                     // Get device address
+                    bool is_write = !(bus->tx_byte & 1);
                     bus->sel_addr = bus->tx_byte >> 1;
-                    if (i2c_oc_get_dev(bus, bus->sel_addr)) bus->status &= ~I2C_OC_SR_ACK;
+                    i2c_dev_t* i2c_dev = i2c_oc_get_dev(bus, bus->sel_addr);
+                    if (i2c_dev && (!i2c_dev->start || i2c_dev->start(i2c_dev, is_write)))
+                        bus->status &= ~I2C_OC_SR_ACK;
                 } else {
                     // Write byte
                     i2c_dev_t* i2c_dev = i2c_oc_get_dev(bus, bus->sel_addr);
diff --git a/src/devices/i2c-oc.h b/src/devices/i2c-oc.h
index e88286c..42902e9 100644
--- a/src/devices/i2c-oc.h
+++ b/src/devices/i2c-oc.h
@@ -31,9 +31,9 @@ typedef struct {
     // Return false on NACK or no data to read
     bool (*write)(void* dev, uint8_t byte);
     bool (*read)(void* dev, uint8_t* byte);
-    //void (*start)(); // Not needed since r/w implicitly starts a transaction?
-    void (*stop)();    // Stop the current transaction
-    void (*remove)();  // Device cleanup
+    bool (*start)(void* dev, bool is_write);
+    void (*stop)(void* dev);    // Stop the current transaction
+    void (*remove)(void* dev);  // Device cleanup
 } i2c_dev_t;
 
 PUBLIC i2c_bus_t* i2c_oc_init(rvvm_machine_t* machine, rvvm_addr_t base_addr, plic_ctx_t* plic, uint32_t irq);
```

**Comment by X547** [comment]
> Don't the actual r/w calls give enough information tho?

Start indicates single read/write group. In theory it can also separate two read or two write groups. Start event also announce information stored in bit 0 of address will be next operations read or write (it must match and be the same).

**Comment by LekKit** [comment]
Implemented in 486a00c. Tell me if there is anything else I might have missed, I'll be back to ~~userland emulation stuff~~ procrastination soon.

For your HID devices, there are some other changes to be made in codebase internally. Namely, public HID API directly maps to PS2 devices as of now, whether the intention was to actually dispatch them indirectly to appropriate device type.
I hope it's not too complicated to make some changes locally first (Perhaps simply get rid of PS2) just to check everything is working.

Btw, if you're too busy working on Haiku (I assume), I can take the responsibility of implementing HID devices as well maybe. At least I'll get to know more, and you can fix stuff afterwards, so tell me if this is desirable.

There is also one questionable thing to be fixed for v0.5 (Usernet TCP buffering), after which I'd like to tag this release and continue HID work without more board breakage per v0.5. The I2C HID, RTL8169, etc will be then aiming 0.6 release (No one cares, but I don't like stretching more features over months of work in staging)

**Comment by X547** [comment]
I made some WIP i2c-hid device driver: https://github.com/X547/RVVM/commit/98725231d2e8a2086cc24fe2571591210ff5a8d8. Not functional for now.

It need more abstraction (generic i2c-hid device and separate keyboard/mouse devices). But I want to make something functional first and refactor later.

**Comment by LekKit** [comment]
> I made some WIP i2c-hid device driver: [X547@9872523](https://github.com/X547/RVVM/commit/98725231d2e8a2086cc24fe2571591210ff5a8d8). Not functional for now.
> 
> It need more abstraction (generic i2c-hid device and separate keyboard/mouse devices). But I want to make something functional first and refactor later.

Very good 😺! Some more points:
- Will there be a way for transparent switch to relative movement? There are applications that lock the guest cursor (i.e. games), and they generally work terrible with absolute input. Ideally, there would be a hotkey in RVVM GUI to lock host cursor, then start reporting "true" relative movement to guest through hid_mouse_move(). I'll see how that plays out...
- I assume since I2C HID spec is born from USB HID, you probably won't need any translation from "RVVM HID API" keycodes
- Should I make some kind of automatic I2C address allocation in i2c-oc? I.e. you pass addr = I2C_ADDR_ANY and it picks any free usable address for your device, then returns it back (This is important, since there are reserved address ranges, etc)
- Should I upstream your changes related to automatic I2C bus attachments (rvvm_get_i2c_bus), etc? Your branch will probably need rebasing anyways before merging (That "default ATA" patch, I don't want to see it upstream. NVMe support should be improved instead, and if any assistance/fixes are needed then this is really important.)
- You seemingly mixed TABs and 4-spaces in your source, which is confusing when viewing patch on GH and in some IDEs (TABs are rendered as 8 spaces, that's why I don't like them - they are inconsistent), don't forget to fix that)

**Comment by X547** [comment]
> * Will there be a way for transparent switch to relative movement? There are applications that lock the guest cursor (i.e. games), and they generally work terrible with absolute input. Ideally, there would be a hotkey in RVVM GUI to lock host cursor, then start reporting "true" relative movement to guest through hid_mouse_move(). I'll see how that plays out...

Probably can be handled with 2 report IDs: one for absolute coordinates and second for relative.

> Should I make some kind of automatic I2C address allocation in i2c-oc? I.e. you pass addr = I2C_ADDR_ANY and it picks any free usable address for your device, then returns it back (This is important, since there are reserved address ranges, etc)

Yes, it will be nice to have automatic i2c device address allocation.

> * Should I upstream your changes related to automatic I2C bus attachments (rvvm_get_i2c_bus), etc? 

Yes if you are fine with that. I2C bus also need to expose FDT node to attach child devices.

> Your branch will probably need rebasing anyways before merging (That "default ATA" patch, I don't want to see it upstream. NVMe support should be improved instead, and if any assistance/fixes are needed then this is really important.)

My fork is not supposed to be directly merged and it can contain my private hacks that are not useful in upstream like switch to ATA. It is supposed to be rebased after upstream updates.

> You seemingly mixed TABs and 4-spaces in your source, which is confusing when viewing patch on GH and in some IDEs (TABs are rendered as 8 spaces, that's why I don't like them - they are inconsistent), don't forget to fix that)

Haiku native code editor Pe support tab indentation only. If you press tab key, it will insert tab character. If you want space indent, you need to type four spaces manually. Anyway it can be cleaned up later.

**Comment by LekKit** [comment]
> Probably can be handled with 2 report IDs: one for absolute coordinates and second for relative.

Fine, as long as it can be treated as a single hid_mouse handle in the outside world

> Yes if you are fine with that. I2C bus also need to expose FDT node to attach child devices.

Something like `struct fdt_node* i2c_bus_get_fdt(i2c_bus_t* bus)` will do? I initially thought the HID FDT nodes would reference the i2c phandle (Like with PLIC), but if they are nested this is the only proper way to do that.

> My fork is not supposed to be directly merged

Ahh I see. Good luck)

**Comment by X547** [comment]
> Something like `struct fdt_node* i2c_bus_get_fdt(i2c_bus_t* bus)` will do?

Yes. This is how I did: https://github.com/X547/RVVM/blob/12d0bf313336c7df3c05b7a79e95f0d5d4f4bec9/src/devices/i2c-oc.c#L244. Choose function name you like.

**Comment by LekKit** [comment]
> > Something like `struct fdt_node* i2c_bus_get_fdt(i2c_bus_t* bus)` will do?
> 
> Yes. This is how I did: https://github.com/X547/RVVM/blob/12d0bf313336c7df3c05b7a79e95f0d5d4f4bec9/src/devices/i2c-oc.c#L244. Choose function name you like.

Btw, you are not required to wrap code handling struct fdt_node in USE_FDT, this structure is forward-declared regardless. Only actual calls to FDT lib should be.
I don't remember why exactly FDT is optional, but a while ago RVVM perfectly ran on microcontrollers, and ditching runtime FDT generation in favor of precompiled one was a good code size saver. Unused devices were simply trimmed by LTO. The more you know)

**Comment by LekKit** [comment]
Upstreamed all the preliminary I2C changes, use `i2c_bus_fdt_node()` (Explicitly defines this is a bus node and not a device)

**Comment by X547** [comment]
I got I2C HID Tablet working in Haiku. No interrupts yet, polling is used.

![screenshot503](https://user-images.githubusercontent.com/45505339/205159653-5342bf03-f5f9-48f5-a591-464e5391df32.png)

[i2c-hid.log](https://github.com/LekKit/RVVM/files/10136198/i2c-hid.log)

Some problems:
* https://github.com/X547/RVVM/commit/5d49bbdb3a0a7bea17d162acc5335ee6df480f95
* HID tablet use scaled coordinates instead of pixels. So `hid_mouse_place` should be changed to use scaled coordinates or HID device should be able to get current screen resolution. https://github.com/X547/RVVM/commit/ed35a32294565662636483ecf3d4849e21e468de

**Comment by LekKit** [comment]
> I got I2C HID Tablet working in Haiku

Amazing! Does it work in Linux? Gonna check that out soon.

> HID tablet use scaled coordinates instead of pixels. 

Oh that's unfortunate, I'm voting for `hid_mouse_screenres(x, y)` but that decision is a complicated one.
I hope it could  work without floats? At least CPU threads (MMIO/I2C handlers, etc) should *never* use floats except for actual emulation, since the FPU state is directly owned by guest as well, and it's pretty much a no-no in the codebase. A while ago I accidentally used floats for vector scaling, and later RV tests started failing... It was a tough debugging, I tell ya.


Btw, writing this to you from inside RVVM (heh), I've been using it in tty/SDL for the last  ~12 hours to see how stable/usable it is. Feels like actually owning a RISC-V device except for the lack of gpu (which led me to use LXDE, but without a compositor everything is pretty smooth). I've some more ideas concerning dynamic JIT cache scaling too, since my default cache size of 16 megs is actually a limiting factor for complex guests.
![image](https://user-images.githubusercontent.com/50500857/205236372-928c3bfd-2213-41ab-8e28-7c830ce5ce17.png)

**Comment by LekKit** [comment]
Running Linux guest under your branch:
```
[    5.683005] i2c_hid_of 0-0008: failed to reset device: -61
[   11.826750] i2c_hid_of 0-0008: failed to reset device: -61
[   17.971236] i2c_hid_of 0-0008: failed to reset device: -61
[   24.115081] i2c_hid_of 0-0008: failed to reset device: -61
[   25.138881] i2c_hid_of 0-0008: can't add hid device: -61
[   25.152375] i2c_hid_of: probe of 0-0008 failed with error -61
```

After using `plic_send_irq(mouse->plic, mouse->irq)` in `mouse_update()`:
```
[    0.765937] input: hid-over-i2c 0001:0001 as /devices/platform/soc/10030000.i2c/i2c-0/0-0008/0018:0001:0001.0001/input/input1
[    0.778111] hid-generic 0018:0001:0001.0001: input: I2C HID v1.00 Mouse [hid-over-i2c 0001:0001] on 0-0008
```

This doesn't give any input tho, looks like Linux doesn't support i2c-hid polling or something is missing.

**Comment by X547** [comment]
> After using `plic_send_irq(mouse->plic, mouse->irq)` in `mouse_update()`:

Note that currently there are no code that reset interrupt pending flag.

**Comment by LekKit** [comment]
> > After using `plic_send_irq(mouse->plic, mouse->irq)` in `mouse_update()`:
> 
> Note that currently there are no code that reset interrupt pending flag.

I noticed that and simply bang IRQs directly to the guest on each update. Hacky, but spurious interrupts should be no problem. Btw, reset should also send an interrupt according to spec.
What is problematic though:
```
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_stop
...
```

Each time I use mouse, the Linux guest receives (And handles) an interrupt, initiates an I2C transaction but doesn't bother to actually read device registers (???)

**Comment by X547** [comment]
> but doesn't bother to actually read device registers (???)

Because max input report size is set to zero (should be 8)?

https://github.com/X547/RVVM/blob/ed35a32294565662636483ecf3d4849e21e468de/src/devices/i2c-hid.c#L106

**Comment by LekKit** [comment]
> > but doesn't bother to actually read device registers (???)
> 
> Because max input report size is set to zero (should be 8)?
> 
> https://github.com/X547/RVVM/blob/ed35a32294565662636483ecf3d4849e21e468de/src/devices/i2c-hid.c#L106

Tried increasing wMaxInputLength and wMaxOutputLength, etc, device fails to reset each time they aren't zero.

<details>
  <summary>Kernel & RVVM log (Warning: Huge)</summary>

```
[    0.570840] i2c_dev: i2c /dev entries driver
[    0.578150] usbcore: registered new interface driver usbhid
[    0.580864] usbhid: USB HID core driver
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 1)
hid_mouse_i2c_write, io_offset: 0, val: 0x1
hid_mouse_i2c_write, io_offset: 1, val: 00
  reg: 1
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x1e
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x1
hid_mouse_i2c_read, io_offset: 4
  val: 0x4a
hid_mouse_i2c_read, io_offset: 5
  val: 00
hid_mouse_i2c_read, io_offset: 6
  val: 0x2
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_read, io_offset: 8
  val: 0x3
hid_mouse_i2c_read, io_offset: 9
  val: 00
hid_mouse_i2c_read, io_offset: 10
  val: 0x8
hid_mouse_i2c_read, io_offset: 11
  val: 00
hid_mouse_i2c_read, io_offset: 12
  val: 0x4
hid_mouse_i2c_read, io_offset: 13
  val: 00
hid_mouse_i2c_read, io_offset: 14
  val: 0x8
hid_mouse_i2c_read, io_offset: 15
  val: 00
hid_mouse_i2c_read, io_offset: 16
  val: 0x5
hid_mouse_i2c_read, io_offset: 17
  val: 00
hid_mouse_i2c_read, io_offset: 18
  val: 0x6
hid_mouse_i2c_read, io_offset: 19
  val: 00
hid_mouse_i2c_read, io_offset: 20
  val: 0xda
hid_mouse_i2c_read, io_offset: 21
  val: 0x9
hid_mouse_i2c_read, io_offset: 22
  val: 0x90
hid_mouse_i2c_read, io_offset: 23
  val: 0x90
hid_mouse_i2c_read, io_offset: 24
  val: 0x1
hid_mouse_i2c_read, io_offset: 25
  val: 00
hid_mouse_i2c_read, io_offset: 26
  val: 00
hid_mouse_i2c_read, io_offset: 27
  val: 00
hid_mouse_i2c_read, io_offset: 28
  val: 00
hid_mouse_i2c_read, io_offset: 29
  val: 00
hid_mouse_i2c_stop
[    0.587088] NET: Registered PF_INET6 protocol family
hid_mouse_i2c_start(is_write: 1)
hid_mouse_i2c_write, io_offset: 0, val: 0x5
hid_mouse_i2c_write, io_offset: 1, val: 00
  reg: 5
hid_mouse_i2c_write, io_offset: 2, val: 00
hid_mouse_i2c_write, io_offset: 3, val: 0x8
  command: 8
hid_mouse_i2c_stop
[    0.599420] Segment Routing with IPv6
[    0.601931] In-situ OAM (IOAM) with IPv6
[    0.604642] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    0.610531] NET: Registered PF_PACKET protocol family
[    0.615511] ALSA device list:
[    0.617628]   No soundcards found.
[    0.654607] input: AT Raw Set 2 keyboard as /devices/platform/soc/20001000.ps2/serio0/input/input0
hid_mouse_i2c_start(is_write: 1)
hid_mouse_i2c_write, io_offset: 0, val: 0x5
hid_mouse_i2c_write, io_offset: 1, val: 00
  reg: 5
hid_mouse_i2c_write, io_offset: 2, val: 00
hid_mouse_i2c_write, io_offset: 3, val: 0x1
  command: 1
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 00
hid_mouse_i2c_read, io_offset: 4
  val: 00
hid_mouse_i2c_read, io_offset: 5
  val: 00
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x7b
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0xe9
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x79
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0xde
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x74
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0xc6
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x72
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0xba
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x70
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0xa5
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x70
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0x9f
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x71
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0x95
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x72
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0x8f
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x74
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0x88
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x75
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0x86
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
hid_mouse_i2c_start(is_write: 0)
hid_mouse_i2c_read, io_offset: 0
  val: 0x8
hid_mouse_i2c_read, io_offset: 1
  val: 00
hid_mouse_i2c_read, io_offset: 2
  val: 00
hid_mouse_i2c_read, io_offset: 3
  val: 0x7b
hid_mouse_i2c_read, io_offset: 4
  val: 0x2
hid_mouse_i2c_read, io_offset: 5
  val: 0x87
hid_mouse_i2c_read, io_offset: 6
  val: 00
hid_mouse_i2c_read, io_offset: 7
  val: 00
hid_mouse_i2c_stop
[    5.709546] i2c_hid_of 0-0008: failed to reset device: -61
```

</details>


Perhaps I should read the spec more, if you have any good third party sources I'll like to see them as well (I2C HID info is pretty scarce and hard to find with unrelated USB stuff showing up)
A good resource for descriptor generation/parsing: https://eleccelerator.com/usbdescreqparser/

**Comment by X547** [comment]
> Tried increasing wMaxInputLength and wMaxOutputLength, etc, device fails to reset each time they aren't zero.

Try to set register to zero on reset:
```
static void hid_mouse_reset(hid_mouse_t* mouse)
{
	mouse->reg = 0;

...

static void hid_mouse_i2c_stop(void* dev)
{
	hid_mouse_t* mouse = (hid_mouse_t*)dev;
	spin_lock(&mouse->lock);
	//fprintf(stderr, "hid_mouse_i2c_stop\n");
	switch (mouse->command) {
	case I2C_HID_COMMAND_RESET:
		hid_mouse_reset(mouse);
		break;
	default:
		mouse->reg = HID_MOUSE_INPUT_REG;
		mouse->command = 0;
		break;
	}
	spin_unlock(&mouse->lock);
}
```

**Comment by LekKit** [comment]
> Try to set register to zero on reset

This unfortunately doesn't help. Btw, before the device probe fails (due to timeout) I can see the kernel at least tries to respond to HID interrupts & reads the device state properly, it's only reset that's broken now

I'll be back after properly reading the spec / kernel driver sources and checking everything is right. If anything, the kernel in question has been posted here, so you may also experiment with it.

Summary so far:
- Idk how interrupt level works, need to read spec first. What I know though, is that sending interrupts unconditionally is a fair temporary solution (I haven't seen line-state based interrupts in other devices yet)
- wMaxInputLength shouldn't be zero if we want to actually handle reports
- Device reset fails when it's not zero though

**Comment by X547** [comment]
I am splitting keyboard/mouse and implementing keyboard HID device right now.

**Comment by X547** [comment]
I2C bus automatic address allocation seems broken:
![screenshot504](https://user-images.githubusercontent.com/45505339/205304270-cb05930a-2266-43e8-b062-8daa6c3973f2.png)

**Comment by X547** [comment]
Keyboard and mouse are working in Haiku. See my branch for code.

![screenshot505](https://user-images.githubusercontent.com/45505339/205310442-1562a067-2703-4d20-9757-58c414c0da45.png)

**Comment by LekKit** [comment]
> Keyboard and mouse are working in Haiku. See my branch for code.

Amazing! I assume in future this can also be easily adapted to USB HID due to how it's abstracted, right?)

As for I2C address allocation - weird, I verified that it works & returns unique addresses. Are you sure you used the correct revision?
![image](https://user-images.githubusercontent.com/50500857/205326720-5df5d835-aa92-4f77-b5c6-2a6c7da95629.png)

**Comment by X547** [comment]
> Amazing! I assume in future this can also be easily adapted to USB HID due to how it's abstracted, right?)

Yes, HID devices should be usable with USB HID controller too. Where `hid_dev_t` definition should be moved? To `hid_api.h`? `hid_dev.h`?

> As for I2C address allocation - weird, I verified that it works & returns unique addresses. Are you sure you used the correct revision?

It seems fixed after I moved `i2c_dev_t` definition to stack variable. I first assumed that `i2c_attach_dev` save pointer to `i2c_dev_t` instance, but it actually save a copy to an array by value.

**Comment by LekKit** [comment]
> Yes, HID devices should be usable with USB HID controller too. Where `hid_dev_t` definition should be moved? To `hid_api.h`? `hid_dev.h`?

Right, hid_dev.h. I even thought of the same name when looking at your commit, huh)
Header hid_api is more or less for outside-world manipulation of input devices, and hid_dev would be for guest-side.

> It seems fixed after I moved `i2c_dev_t` definition to stack variable. I first assumed that `i2c_attach_dev` save pointer to `i2c_dev_t` instance, but it actually save a copy to an array by value.

Yes, since it should be able to modify the `addr`.
It should also allow to optimize device lookup without breaking the API, and removes the need to maintain this description somewhere else.

**Comment by X547** [comment]
Race condition and random crash fix (on Haiku input event handling is done in another thread):

``` diff
diff --git a/src/devices/fb_window.c b/src/devices/fb_window.c
index 39d192e..2e3fbc7 100644
--- a/src/devices/fb_window.c
+++ b/src/devices/fb_window.c
@@ -122,14 +122,14 @@ bool fb_window_init_auto(rvvm_machine_t* machine, uint32_t width, uint32_t heigh
     window->fb.width = width;
     window->fb.height = height;
     window->fb.format = RGB_FMT_A8R8G8B8;
+    window->machine = machine;
+    window->keyboard = hid_keyboard_init_auto(machine);
+    window->mouse = hid_mouse_init_auto(machine);
     if (!fb_window_create(window)) {
         rvvm_error("Window creation failed");
         free(window);
         return false;
     }
-    window->machine = machine;
-    window->keyboard = hid_keyboard_init_auto(machine);
-    window->mouse = hid_mouse_init_auto(machine);
     
     framebuffer_init_auto(machine, &window->fb);
     
```

**Comment by X547** [comment]
Updated source code. Does it work now with Linux?

**Comment by LekKit** [comment]
> Race condition and random crash fix

Thanks; I'll also document that fb_window_create() could've additional side effects other than creating a window.
One problem though is, if window creation fails (Common on Linux when I'm running from ssh, etc) the guest now has useless input devices which I'd like to not create.
Also note that public HID API should be able to handle NULL pointers and ignore such calls (This is a common error handling strategy in RVVM for many types of handles, I.e. we may open a file and then pass returned NULL handle to rvread(), it should return an error then)

Is there a way to run ThreadSanitizer on Haiku? That's how I usually catch such problems, but for now I only do that on Linux, i.e. all other platform specific code is left without sanitizer coverage.

**Comment by LekKit** [comment]
> Updated source code. Does it work now with Linux?

Gonna test that soon

**Comment by X547** [comment]
> Is there a way to run ThreadSanitizer on Haiku?

Not yet, but I managed to get Clang/LLD 15 on Haiku. It seems to need port some runtime libraries.

**Comment by LekKit** [comment]
> Updated source code. Does it work now with Linux?

It works after commenting out `if (!i2c_hid->int_pending) {` in `i2c_hid_input_available()`. Something somewhere is not resetting this flag properly.
In all other regards this is great, some warnings here & there which I'll either fix after merge or you will fix when refactoring.

What's left though is a decision what to do with `hid_mouse_place()`. I think it will be a good idea to add `hid_mouse_resolution()` which will be called to set size of the workplace, then window systems won't need any changes (And if they wish to support dynamic resolution, they could use this API at any moment).
Do you support this? I want the public API to be clean, and not force users to calculate scaled coordinated on their own.

**Comment by LekKit** [comment]
> Something somewhere is not resetting this flag properly.

@X547 The problem is here
https://github.com/X547/RVVM/blob/fa9ccac90dd81e6396c21f65c61d529b6200c97e/src/devices/i2c-hid.c#L147

It sets `int_pending` after hid reset but no one clears it afterwards. Only input handlers clear it.
Udp: Sorry, accidentally copied wrong line link.

**Comment by X547** [comment]
> It sets `int_pending` after hid reset but no one clears it afterwards. Only input handlers clear it.
> Udp: Sorry, accidentally copied wrong line link.

I2C HID spec:
> 7.4 INTERRUPT LINE MANAGEMENT
> DEVICE is responsible to assert the interrupt line (level trigger interrupt required) when it has data.
> DEVICE must keep the interrupt line asserted until the data has been fully read by the HOST. After this
> point, the DEVICE must de-assert the interrupt line if it has no more data to report to the HOST.
> When the DEVICE has new data to report, it must repeat the process above.

I2C HID use level triggered interrupt, but RVVM PLIC API supports only edge-triggered interrupts.

So correct behavior should be asserting interrupt line when report ID queue is not empty or when reset confirmation is not yet read.

**Comment by X547** [comment]
> @X547 The problem is here
> https://github.com/X547/RVVM/blob/fa9ccac90dd81e6396c21f65c61d529b6200c97e/src/devices/i2c-hid.c#L147

https://github.com/X547/RVVM/commit/f8205499f5e6dd720e536fe5565639798aa61a83

**Comment by LekKit** [comment]
> I2C HID use level triggered interrupt, but RVVM PLIC API supports only edge-triggered interrupts.

It's a mix of both, if you do `plic_send_irq()` twice it will not send 2 interrupts at the same time. Resetting the IRQ is done by the kernel though and not by the device. That's why I'm not even sure that `int_pending` flag is needed, since the same thing is maintained inside the PLIC.

As for true level-based interrupts API, I will implement it in future probably. However no devices needed that yet (If the IRQ handler fails to process device state, the device could simply re-send IRQ and it would look same. Other scenarios usually indicate a buggy guest.).

**Comment by X547** [comment]
> It's a mix of both, if you do `plic_send_irq()` twice it will not send 2 interrupts at the same time. Resetting the IRQ is done by the kernel though and not by the device.

TinyEMU PLIC allow to assert/deassert interrupt: https://github.com/X547/TinyEMU/blob/master/riscv_machine.c#L327.

**Comment by LekKit** [comment]
> > It's a mix of both, if you do `plic_send_irq()` twice it will not send 2 interrupts at the same time. Resetting the IRQ is done by the kernel though and not by the device.
> 
> TinyEMU PLIC allow to assert/deassert interrupt: https://github.com/X547/TinyEMU/blob/master/riscv_machine.c#L327.

I know, the RVVM PLIC API simply is built around the fact that most device interrupts are message-based. You can ask @cerg2010cerg2010 for more clues)

**Comment by X547** [comment]
> That's why I'm not even sure that `int_pending` flag is needed, since the same thing is maintained inside the PLIC.

I considered a case when interrupt is enabled but nobody actually read input report (or read it after/slowly etc.). Without `int_pending` flag it will send meaningless interrupt in every mouse move etc..

**Comment by LekKit** [comment]
> I considered a case when interrupt is enabled but nobody actually read input report. Without `int_pending` flag it will send meaningless interrupt in every mouse move etc..

Aight then, most devices have IRQ enable bits for similar thing (See `i2c-oc.c@74`, etc). I just thought you're emulating line interrupts behavior by that, which isn't needed.

What about the decision for `hid_mouse_place()`?

**Comment by X547** [comment]
> Aight then, most devices have IRQ enable bits for similar thing (See i2c-oc.c@74, etc).

Does it work after fix?

Real I2C HID device have no such flag and it depends on GPIO controller. GPIO controller probably have interrupt pending flag.

RISC-V PLIC also supports level-triggered interrupts directly:

> If the global interrupt source uses level-sensitive interrupts, the gateway will convert the first assertion of the interrupt level into an interrupt request, but thereafter the gateway will not forward an additional interrupt request until it receives an interrupt completion message. On receiving an interrupt completion message, **if the interrupt is level-triggered and the interrupt is still asserted, a new interrupt request will be forwarded to the PLIC core**. The gateway does not have the facility to retract an interrupt request once forwarded to the PLIC core. If a level-sensitive interrupt source deasserts the interrupt after the PLIC core accepts the request and before the interrupt is serviced, the interrupt request remains present in the IP bit of the PLIC core and will be serviced by a handler, which will then have to determine that the interrupt device no longer requires service.

https://github.com/riscv/riscv-plic-spec/blob/master/riscv-plic.adoc#interrupt-gateways

> What about the decision for `hid_mouse_place()`?

`hid_mouse_resolution()` seems fine to me. You can push changes and I will rebase and adapt my branch.

**Comment by LekKit** [comment]
Sadly Linux guests ignore `HID_KEY_POWER` which is what I hoped to work to cover #60...

**Comment by LekKit** [comment]
> RISC-V PLIC also supports level-triggered interrupts directly:

I see (And knew about that spec), I will implement such thing in near future (PLIC refactoring is also needed). The fact is, current API is still usable if devices care about line interrupts, but the responsibility to re-send the interrupt lies on them.

There is also APLIC on the horizon... See https://github.com/riscv/riscv-aia

**Comment by LekKit** [comment]
Implemented `hid_mouse_resolution()` in staging, it's called each time before window creation

**Comment by X547** [comment]
Also what we should do with PS/2 input devices? Currently I add `_ps2` suffix to avoid symbol conflict, but it makes PS/2 input devices a dead code.

**Comment by LekKit** [comment]
> Also what we should do with PS/2 input devices? Currently I add `_ps2` suffix to avoid symbol conflict, but it makes PS/2 input devices a dead code.

That's fine, PS2 joins the **Land of the Cursed Devices** and will wait before I rewrite HID layer abstractions for librvvm usage (Explicit creation of PS2 devices).
I should only verify that Ubuntu & other generic distros work with I2C HID (They most likely do), this doesn't break anything else because PS2 didn't work elsewhere anyways.

Some people with custom kernel configs will probably curse me, but we have to do it at some point.

That said, I'll need to verify your code under lots of sanitizers, analyzers, fuzzing, etc before 0.5 release. Hope you also can help with `tap_user` TCP buffering because I'm stuck and this is one of v0.5 blockers.

**Comment by X547** [comment]
> Hope you also can help with `tap_user` TCP buffering because I'm stuck and this is one of v0.5 blockers.

I don't see an open issue for that.

**Comment by LekKit** [comment]
Ubuntu seems fine with I2C HID. It warns about this, however:
`i2c_hid_of 0-0008: supply vdd not found, using dummy regulator`

I assume this can be safely ignored

**Comment by X547** [comment]
> I assume this can be safely ignored

It should be optional FDT parameter:

> Additional optional properties:
>
> Some devices may support additional optional properties to help with, e.g.,
> power sequencing. The following properties can be supported by one or more
> device-specific compatible properties, which should be used in addition to the
> "hid-over-i2c" string.
> 
> - compatible:
>   * "wacom,w9013" (Wacom W9013 digitizer). Supports:
>     - vdd-supply (3.3V)
>     - vddl-supply (1.8V)
>     - post-power-on-delay-ms
> 
https://www.kernel.org/doc/Documentation/devicetree/bindings/input/hid-over-i2c.txt

**Comment by LekKit** [comment]
> I don't see an open issue for that.

I simply don't know how to implement proper TCP window. I thought I'll do it at some point and then was busy with other stuff.
Right now userspace networking just blows TCP segments at the guest and ignores ACKs, if it doesn't receive them in time the connection hangs. This is very bad, but my attempts to implement proper buffering made the performance abysmal.

**Comment by X547** [comment]
> Implemented `hid_mouse_resolution()` in staging, it's called each time before window creation

Updated my branch.

**Comment by LekKit** [comment]
Bad news: Linux guests do not seem to care that this is an absolute input device....

Online sources suggest using digitizer HID class instead of mouse

Will inspect other sources and report back

**Comment by X547** [comment]
Created pull requests, you can add code review there (bugs, bad design, code style violations etc.).

**Comment by LekKit** [comment]
> Bad news: Linux guests do not seem to care that this is an absolute input device....

Seems to depend on X configuration or something though, LXDE supports absolute cursor properly

**Comment by X547** [comment]
> Bad news: Linux guests do not seem to care that this is an absolute input device....

I added tablet HID report descriptor from QEmu, I suppose that it is heavily tested.

**Comment by LekKit** [comment]
> > Bad news: Linux guests do not seem to care that this is an absolute input device....
> 
> I added tablet HID report descriptor from QEmu, I suppose that it is heavily tested.

It's Buildroot Xorg problem then. GPM (TTY-based Linux cursor) is also only working with relative movement.

**Comment by X547** [comment]
Seems actual deadlock:
```
WARN: Possible deadlock at /Haiku/data/packages/RVVM/src/devices/hid-mouse.c@97
WARN: The lock was previously held at /Haiku/data/packages/RVVM/src/devices/hid-mouse.c@213
WARN: Version: RVVM v0.5
WARN: Attempting to recover execution...
```
a->b, b->a deadlock pattern with I2C HID controller and HID device locks.

**Comment by LekKit** [comment]
Yes I've seen this as well. Weird how fuzzing didn't catch this, huh.
```
WARN: Possible deadlock at src/devices/hid-mouse.c@97
WARN: The lock was previously held at src/devices/hid-mouse.c@213
WARN: Version: RVVM v0.5-49249a3-git
WARN: Attempting to recover execution...
 * * * * * * *

WARN: Possible deadlock at src/devices/i2c-hid.c@155
WARN: The lock was previously held at src/devices/i2c-hid.c@318
WARN: Version: RVVM v0.5-49249a3-git
WARN: Attempting to recover execution...
 * * * * * * *

WARN: Possible deadlock at src/devices/hid-keyboard.c@109
WARN: The lock was previously held at src/devices/hid-keyboard.c@192
WARN: Version: RVVM v0.5-49249a3-git
WARN: Attempting to recover execution...
 * * * * * * *
```

Sounds very similar to the old PS2 issue, which also brought the idea of lock debugging (It took ages for us to understand why it deadlocks before that). CPU thread could've held the controller lock while the GUI thread would take the HID lock, then they deadlock waiting for each other.
@cerg2010cerg2010 resolved this in PS2 by using a single lock per device & controller, but this is hacky from API standpoint and I think there is a better way.

My thoughts: locks on the device side should be held only when manipulating on device data, not when we are actually notifying back HID/I2C buses. I.e. something like this:
```c
spin_lock(&mouse->lock);
mouse->x = ....;
spin_unlock(&mouse->lock);
mouse->hid.has_input(...);
```

**Comment by X547** [comment]
> My thoughts: locks on the device side should be held only when manipulating on device data, not when we are actually notifying back HID/I2C buses. I.e. something like this:

https://github.com/LekKit/RVVM/pull/67

**Comment by LekKit** [comment]
Will be reopened if any more issues are spotted.
For now, everything seems fine.
